components/locale/src/langid.rs

```diff
@@ -0,0 +1,138 @@
+use std::fmt::Write;
```

Comment:
I have couple of problems with Language ID (not Locale ID):

1. Having 2 similar APIs (Locale vs Language) will be confusing - I know they can "upgrade" from one to other, still.
2. Ecma402 doesn't have Language Id, and I am not seeing demand for it
3. It feels like a low lever struct, useful for resource loading (CLDR) that doesn't have a place in public surface

If we need it for internals, we can add it as private struct, in multiple shapes for optimization (with 0, 1, * variants)...

WDYT?

Reply:
> Having 2 similar APIs (Locale vs Language) will be confusing - I know they can "upgrade" from one to other, still.

I don't understand the concern. LDML has `Language Identifier` well specified and defined.

> Ecma402 doesn't have Language Id, and I am not seeing demand for it

I must say I really don't like the fixation on ECMA402. I understand we want to supplement ECMA402 needs, but I see increased arguments against implementation of Unicode Standard because "ECMA402 doesn't expose it".

> It feels like a low lever struct, useful for resource loading (CLDR) that doesn't have a place in public surface

The structure is used as the base language negotiation API internally by Gecko. I hope it warrants justification as a use case.

It's not "internals". It's an implementation of [Unicode Language Identifier ](https://unicode.org/reports/tr35/#Unicode_language_identifier) standard.

I also want to point out, that ECMA402 implements a subset of what software needs. Majority of ICU4X customers need more than ECMA402 and the tradeoffs between API surface and features is very different between Web Standard and library API.

Reply:
FWIW, I like the idea of exposing the individual pieces (Language, Script, Region, Variant, Variants) and allowing clients to implement their own struct with the subset most relevant for them.

> It's not "internals". It's an implementation of Unicode Language Identifier standard.

Yes, but what is the use case for that standard?  We don't need to expose API for every single Unicode standard just because it exists.

> The structure is used as the base language negotiation API internally by Gecko.

Why does Gecko need LanguageIdentifier for this use case?  Why don't you use Locale instead?

> I also want to point out, that ECMA402 implements a subset of what software needs. Majority of ICU4X customers need more than ECMA402 and the tradeoffs between API surface and features is very different between Web Standard and library API.

Discussed offline.  We should be more explicit in the charter about the use cases that warrant expansion of functional units beyond the scope of 402.

Reply:
> Yes, but what is the use case for that standard? We don't need to expose API for every single Unicode standard just because it exists.

I don't think we do. As with all argument `ad extremum` :)

> Why does Gecko need LanguageIdentifier for this use case? Why don't you use Locale instead?

Performance and memory considerations mainly.

> Discussed offline. We should be more explicit in the charter about the use cases that warrant expansion of functional units beyond the scope of 402.

Yep!

Reply:
> Performance and memory considerations mainly.

Why can't we design Locale such that it automatically optimizes itself when things like Unicode extension keys are not being used?

Exposing LanguageIdentifier gives a hint, and I would venture to say a ***strong*** hint, to clients that maybe this is something that they should be using, when in reality, clients should almost always be using Locale by default.  Developers new to the field know language and region subtags, like `"en-US"`, and maybe they know script subtags, like `"sr-Cyrl"`, but knowledge of the importance of Unicode extension keywords is much more limited.  If the gold standard i18n library gives them a shiny-looking LangaugeIdentifier type, touting its performance, developers might think they should use it.

Ultimately, clients should use something different only in cases where they have specific performance needs, and those needs are unique to each client.  In my opinion, I'm not convinced by the argument that a "one size fits all" type for performance-optimized language negotiation justifies the inclusion of a new API into the library.

P.S. If the goal is performance and memory considerations for language negotiations, then I should point out that with the exception of [ICU-20997](https://unicode-org.atlassian.net/browse/ICU-20997), ICU LocaleMatcher uses only a fixed-size "language + script + region" type, or LSR, for language negotiation.

Reply:
> Why can't we design Locale such that it automatically optimizes itself when things like Unicode extension keys are not being used?

We may get there. I want to start small and build on top. `LanguageIdentifier` seems like a good, self sufficient, scoped out useful API that can already provide good functionality - as I said, it's used in production Firefox on hundreds of millions of installations and serves us really well.

`Locale` is so far slightly slower, uses more memory and I since it's more complex, there's a larger surface of potential bugs in extensions implementation.

As I said, we have a lot of time to fine-tune the API and decide what we end up with. Until `1.0`, we can easily fold `LanguageIdentifier` into `Locale` once we have `Locale` with satisfying params.

> P.S. If the goal is performance and memory considerations for language negotiations, then I should point out that with the exception of ICU-20997, ICU LocaleMatcher uses only a fixed-size "language + script + region" type, or LSR, for language negotiation.

Gecko language negotiation uses variants as one of the input. We serve separate japanese localization for MacOS as `ja-JP-macos`.

Reply:
I am satisfied with your and Shanes discussion that followed, but still want to clarify my points (I feel some were taken too strongly).

> > Having 2 similar APIs (Locale vs Language) will be confusing - I know they can "upgrade" from one to other, still.
> 
> I don't understand the concern. LDML has `Language Identifier` well specified and defined.
> 
> > Ecma402 doesn't have Language Id, and I am not seeing demand for it
> 
> I must say I really don't like the fixation on ECMA402. I understand we want to supplement ECMA402 needs, but I see increased arguments against implementation of Unicode Standard because "ECMA402 doesn't expose it".

I think fixation is too strong of a word - it was a minor point derived from our overall agenda. I expect us to breach borders, with a good reason, for example implementing language matching which is not a public ES402 API.

> > It feels like a low lever struct, useful for resource loading (CLDR) that doesn't have a place in public surface
> 
> The structure is used as the base language negotiation API internally by Gecko. I hope it warrants justification as a use case.
> 
> It's not "internals". It's an implementation of [Unicode Language Identifier ](https://unicode.org/reports/tr35/#Unicode_language_identifier) standard.

When I said internals, I didn't refer to tr35 where it's public, I was looking at potential use cases in actual products, or our library.

Reply:
> I think fixation is too strong of a word - it was a minor point derived from our overall agenda.

I apologize. My reaction was to your comment, but it wasn't directed at it. I've seen a number of comments about the design of ICU4X that were using ECMA402 as a qualifier of whether it belongs or not, and I reacted to that. It was unwarranted and unnecessary. I'm sorry.